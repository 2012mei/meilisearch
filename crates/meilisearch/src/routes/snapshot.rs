use actix_web::web::Data;
use actix_web::{web, HttpRequest, HttpResponse};
use index_scheduler::IndexScheduler;
use meilisearch_types::error::ResponseError;
use meilisearch_types::tasks::KindWithContent;
use tracing::debug;

use crate::analytics::Analytics;
use crate::extractors::authentication::policies::*;
use crate::extractors::authentication::GuardedData;
use crate::extractors::sequential_extractor::SeqHandler;
use crate::routes::{get_task_id, is_dry_run, SummarizedTaskView};
use crate::Opt;

pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(web::resource("").route(web::post().to(SeqHandler(create_snapshot))));
}

crate::empty_analytics!(SnapshotAnalytics, "Snapshot Created");

pub async fn create_snapshot(
    index_scheduler: GuardedData<ActionPolicy<{ actions::SNAPSHOTS_CREATE }>, Data<IndexScheduler>>,
    req: HttpRequest,
    opt: web::Data<Opt>,
    analytics: web::Data<Analytics>,
) -> Result<HttpResponse, ResponseError> {
    analytics.publish(SnapshotAnalytics::default(), &req);

    let task = KindWithContent::SnapshotCreation;
    let uid = get_task_id(&req, &opt)?;
    let dry_run = is_dry_run(&req, &opt)?;
    let task: SummarizedTaskView =
        tokio::task::spawn_blocking(move || index_scheduler.register(task, uid, dry_run))
            .await??
            .into();

    debug!(returns = ?task, "Create snapshot");
    Ok(HttpResponse::Accepted().json(task))
}
