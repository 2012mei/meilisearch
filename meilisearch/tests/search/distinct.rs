use meili_snap::snapshot;
use once_cell::sync::Lazy;
use serde_json::{json, Value};

use crate::common::Server;

pub(self) static DOCUMENTS: Lazy<Value> = Lazy::new(|| {
    json!([
        {"productId": 1, "shopId": 1},
        {"productId": 2, "shopId": 1},
        {"productId": 3, "shopId": 2},
        {"productId": 4, "shopId": 2},
        {"productId": 5, "shopId": 3},
        {"productId": 6, "shopId": 3},
        {"productId": 7, "shopId": 4},
        {"productId": 8, "shopId": 4},
        {"productId": 9, "shopId": 5},
        {"productId": 10, "shopId": 5}
    ])
});

pub(self) static DOCUMENT_PRIMARY_KEY: &str = "productId";
pub(self) static DOCUMENT_DISTINCT_KEY: &str = "shopId";

/// testing: https://github.com/meilisearch/meilisearch/issues/4078
#[actix_rt::test]
async fn distinct_search_with_offset_no_ranking() {
    let server = Server::new().await;
    let index = server.index("test");

    let documents = DOCUMENTS.clone();
    index.add_documents(documents, Some(DOCUMENT_PRIMARY_KEY)).await;
    index.update_distinct_attribute(json!(DOCUMENT_DISTINCT_KEY)).await;
    index.wait_task(1).await;

    fn get_hits(response: Value) -> Vec<i64> {
        let hits_array = response["hits"].as_array().unwrap();
        hits_array.iter().map(|h| h[DOCUMENT_DISTINCT_KEY].as_i64().unwrap()).collect::<Vec<_>>()
    }

    let (response, code) = index.search_post(json!({"limit": 2, "offset": 0})).await;
    let hits = get_hits(response);
    snapshot!(code, @"200 OK");
    snapshot!(hits.len(), @"2");
    snapshot!(format!("{:?}", hits), @"[1, 2]");

    let (response, code) = index.search_post(json!({"limit": 2, "offset": 2})).await;
    let hits = get_hits(response);
    snapshot!(code, @"200 OK");
    snapshot!(hits.len(), @"2");
    snapshot!(format!("{:?}", hits), @"[3, 4]");

    let (response, code) = index.search_post(json!({"limit": 10, "offset": 4})).await;
    let hits = get_hits(response);
    snapshot!(code, @"200 OK");
    snapshot!(hits.len(), @"1");
    snapshot!(format!("{:?}", hits), @"[5]");

    let (response, code) = index.search_post(json!({"limit": 10, "offset": 5})).await;
    let hits = get_hits(response);
    snapshot!(code, @"200 OK");
    snapshot!(hits.len(), @"0");
}
