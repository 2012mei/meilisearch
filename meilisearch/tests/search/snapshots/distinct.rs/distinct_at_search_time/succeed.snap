---
source: meilisearch/tests/search/distinct.rs
---
{
  "uid": 1,
  "indexUid": "tamo",
  "status": "succeeded",
  "type": "settingsUpdate",
  "canceledBy": null,
  "details": {
    "filterableAttributes": [
      "color.main"
    ]
  },
  "error": null,
  "duration": "[duration]",
  "enqueuedAt": "[date]",
  "startedAt": "[date]",
  "finishedAt": "[date]"
}
